### PR TITLE
MOTECH-2342 Add multiple address lines to OpenMRS Create Patient task action

### PR DESCRIPTION
**Pull request by martokarski**
All fields from OpenMRS REST Web Service documentation (https://wiki.openmrs.org/display/docs/REST+Web+Service+Resources+in+OpenMRS+1.9#RESTWebServiceResourcesinOpenMRS1.9-PersonAddress) are supported now.

### CODE REVIEW

File: openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/impl/OpenMRSActionProxyServiceImpl.java [hunk 1]
```diff
@@ -69,11 +73,25 @@ public void createPatient(String givenName, String middleName, String familyName
...
+        personAddress.setCountyDistrict(countyDistrict);
+        personAddress.setLatitude(latitude);
+        personAddress.setLongitude(longitude);
+        personAddress.setStartDate(Objects.nonNull(startDate) ? startDate.toDate() : null);
+        personAddress.setEndDate(Objects.nonNull(endDate) ? endDate.toDate() : null);
         person.setPreferredAddress(personAddress);
         person.setAddresses(Collections.singletonList(personAddress));
 
-        person.setBirthdate(birthdate.toDate());
+        person.setBirthdate(Objects.nonNull(startDate) ? startDate.toDate() : null);
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/impl/OpenMRSActionProxyServiceImpl.java]: birthdate

File: openmrs-19/src/test/java/org/motechproject/openmrs19/tasks/impl/OpenMRSActionProxyServiceTest.java [hunk 2]
```diff
@@ -111,9 +111,16 @@ public void shouldCreatePatientWithGivenParameters() {
         doReturn(Collections.singletonList(location)).when(locationService).getLocations(location.getName());
 
         openMRSActionProxyService.createPatient(person.getPreferredName().getGivenName(), person.getPreferredName().getMiddleName(),
-                person.getPreferredName().getFamilyName(), person.getPreferredAddress().getAddress1(), new DateTime(person.getBirthdate()),
-                person.getBirthdateEstimated(), person.getGender(), person.getDead(), causeOfDeath.getUuid(), patient.getMotechId(),
-                location.getName(), identifiersMap);
+                person.getPreferredName().getFamilyName(), person.getPreferredAddress().getAddress1(),
```
Comment (kmadej) [inline openmrs-19/src/test/java/org/motechproject/openmrs19/tasks/impl/OpenMRSActionProxyServiceTest.java]: to simplify this code, maybe use new object to get PreferredAddress and then use getters on it

File: openmrs-19/src/test/java/org/motechproject/openmrs19/tasks/impl/OpenMRSActionProxyServiceTest.java [hunk 3]
```diff
@@ -111,9 +111,16 @@ public void shouldCreatePatientWithGivenParameters() {
...
+                person.getPreferredAddress().getAddress2(), person.getPreferredAddress().getAddress3(),
+                person.getPreferredAddress().getAddress4(), person.getPreferredAddress().getAddress5(),
```
Comment (kmadej) [inline openmrs-19/src/test/java/org/motechproject/openmrs19/tasks/impl/OpenMRSActionProxyServiceTest.java]: you are testing this, however I don't see adding these values to the test person

File: openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/impl/OpenMRSActionProxyServiceImpl.java [hunk 4]
```diff
@@ -69,11 +73,25 @@ public void createPatient(String givenName, String middleName, String familyName
         person.setNames(Collections.singletonList(personName));
 
         Person.Address personAddress = new Person.Address();
-        personAddress.setAddress1(address);
+        personAddress.setAddress1(address1);
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/impl/OpenMRSActionProxyServiceImpl.java]: can you add a constructor for address with all these fields

**Comment by kmadej** [comment]
You have to clean up this branch, because there are commits which aren't yours.

Please write an unit test to PersonResourceImplTest, include in test json all this fields

Also add an integration test to MRSPersonServiceIT, it should check if all this field values are properly stored to OpenMRS server.

Lastly, can you check briefly if there are any wrong written test in two mentioned classes.

Remember to check locally if OpenMRSITs pass.

File: openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/impl/OpenMRSActionProxyServiceImpl.java [hunk 5]
```diff
@@ -69,11 +73,25 @@ public void createPatient(String givenName, String middleName, String familyName
...
+        personAddress.setAddress2(address2);
+        personAddress.setAddress3(address3);
+        personAddress.setAddress4(address4);
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/tasks/impl/OpenMRSActionProxyServiceImpl.java]: use constructor

File: openmrs-19/src/test/java/org/motechproject/openmrs19/it/MRSPersonServiceIT.java [hunk 6]
```diff
@@ -46,30 +50,60 @@
...
+        Person fetchedPerson = personAdapter.getPersonByUuid(DEFAULT_CONFIG_NAME, createdPerson.getUuid());
+
+        assertEquals(person.getPreferredName().getGivenName(), fetchedPerson.getPreferredName().getGivenName());
+        assertEquals(person.getPreferredName().getMiddleName(), fetchedPerson.getPreferredName().getMiddleName());
+        assertEquals(person.getPreferredName().getFamilyName(), fetchedPerson.getPreferredName().getFamilyName());
+
+        Person.Address personAddress = person.getPreferredAddress();
+        Person.Address fetchedPersonAddress = fetchedPerson.getPreferredAddress();
+
+        assertEquals(personAddress.getAddress1(), fetchedPersonAddress.getAddress1());
```
Comment (kmadej) [inline openmrs-19/src/test/java/org/motechproject/openmrs19/it/MRSPersonServiceIT.java]: assert two address objects

File: openmrs-19/src/main/java/org/motechproject/openmrs19/domain/Person.java [hunk 7]
```diff
@@ -274,14 +296,14 @@ public boolean equals(Object obj) {
...
 
-            return  Objects.equals(this.uuid, other.uuid) && Objects.equals(this.address1, other.address1) &&
-                    Objects.equals(this.address2, other.address2) && Objects.equals(this.cityVillage, other.cityVillage) &&
-                    Objects.equals(this.stateProvince, other.stateProvince) && Objects.equals(this.country, other.country) &&
-                    Objects.equals(this.postalCode, other.postalCode) && Objects.equals(this.countyDistrict, other.countyDistrict) &&
-                    Objects.equals(this.address3, other.address3) && Objects.equals(this.address4, other.address4) &&
-                    Objects.equals(this.address5, other.address5) && Objects.equals(this.address6, other.address6) &&
-                    Objects.equals(this.startDate, other.startDate) && Objects.equals(this.endDate, other.endDate) &&
-                    Objects.equals(this.latitude, other.latitude) && Objects.equals(this.longitude, other.longitude);
+            return  Objects.equals(this.address1, other.address1) && Objects.equals(this.address2, other.address2) &&
```
Comment (kmadej) [inline openmrs-19/src/main/java/org/motechproject/openmrs19/domain/Person.java:299]: remove uuid from hashcode and please write a brief comment on PR why you remove this

File: openmrs-19/src/test/java/org/motechproject/openmrs19/resource/impl/PersonResourceImplTest.java [hunk 8]
```diff
@@ -34,6 +34,7 @@
     private static final String PERSON_NAME_UPDATE_JSON = "json/person-name-update.json";
     private static final String PERSON_RESPONSE_JSON = "json/person-response.json";
     private static final String CREATE_PERSON_JSON = "json/person-create.json";
+    private static final String UPDATE_ADDRESS_JSON = "json/address-update.json";
```
Comment (kmadej) [inline openmrs-19/src/test/java/org/motechproject/openmrs19/resource/impl/PersonResourceImplTest.java]: please change json file to person-address-update.json and adjust constant name

**Comment by martokarski** [comment]
If we want to compare local copy of Person.Address with fetched, they are different. The difference is uuid field, beacuse it is set only when fetched from server. So, I remove uuid field from Person.Address equals() method.

**Pull request merged**